transports/tls-quic/src/verifier.rs

```diff
@@ -0,0 +1,211 @@
+        presented_certs: &[Certificate],
+        _dns_name: webpki::DNSNameRef<'_>,
+        _ocsp_response: &[u8],
+    ) -> Result<ServerCertVerified, TLSError> {
+        verify_certs(presented_certs).map(|_| ServerCertVerified::assertion())
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        _message: &[u8],
+        _cert: &Certificate,
+        _dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, TLSError> {
+        // The libp2p handshake uses TLS 1.3 (and higher).
+        // Endpoints MUST NOT negotiate lower TLS versions.
+        Err(TLSError::PeerIncompatibleError("Only TLS 1.3 certificates are supported".to_string()))
+    }
+
+    fn verify_tls13_signature(
+        &self,
+        message: &[u8],
+        cert: &Certificate,
+        dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, TLSError> {
+        verify_tls13_signature(cert, dss.scheme, message, dss.sig.0.as_ref())
+    }
+
+    fn supported_verify_schemes(&self) -> Vec<SignatureScheme> {
+        Self::verification_schemes()
+    }
+}
+
+impl rustls::ClientCertVerifier for Libp2pCertificateVerifier {
+    fn offer_client_auth(&self) -> bool {
+        true
+    }
+
+    fn client_auth_root_subjects(
+        &self,
+        _dns_name: Option<&webpki::DNSName>,
+    ) -> Option<DistinguishedNames> {
+        Some(vec![])
+    }
+
+    fn verify_client_cert(
+        &self,
+        presented_certs: &[Certificate],
+        _dns_name: Option<&webpki::DNSName>,
+    ) -> Result<ClientCertVerified, TLSError> {
+        verify_certs(presented_certs).map(|_| ClientCertVerified::assertion())
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        _message: &[u8],
+        _cert: &Certificate,
+        _dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, TLSError> {
+        // The libp2p handshake uses TLS 1.3 (and higher).
+        // Endpoints MUST NOT negotiate lower TLS versions.
+        Err(TLSError::PeerIncompatibleError("Only TLS 1.3 certificates are supported".to_string()))
+    }
+
+    fn verify_tls13_signature(
+        &self,
+        message: &[u8],
+        cert: &Certificate,
+        dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, TLSError> {
+        verify_tls13_signature(cert, dss.scheme, message, dss.sig.0.as_ref())
+    }
+
+    fn supported_verify_schemes(&self) -> Vec<SignatureScheme> {
+        Self::verification_schemes()
+    }
+}
+
+/// When receiving the certificate chain, an endpoint
+/// MUST check these conditions and abort the connection attempt if
+/// (a) the presented certificate is not yet valid, OR
+/// (b) if it is expired.
+/// Endpoints MUST abort the connection attempt if more than one certificate is received,
+/// or if the certificate’s self-signature is not valid.
+fn verify_certs(presented_certs: &[Certificate]) -> Result<(), TLSError> {
+    if presented_certs.len() != 1 {
+        return Err(TLSError::NoCertificatesPresented);
```

Comment:
Confusing error if more than 1 cers presented. Also no explanation of why we expect only 1 certificate

Reply:
The explanation is in the comment to this function: `/// Endpoints MUST abort the connection attempt if more than one certificate is received,` This is a requirement from the libp2p-tls specs.

I know this is a confusing error, however there is no such TLSError to tell the cert chain is too long.